File: app/src/main/java/de/dreier/mytargets/views/selector/ArrowSelector.java
/*
 * Copyright (C) 2017 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.views.selector;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;

import java.util.List;

import de.dreier.mytargets.features.arrows.ArrowListActivity;
import de.dreier.mytargets.features.arrows.EditArrowFragment;
import de.dreier.mytargets.shared.models.db.Arrow;
import de.dreier.mytargets.utils.IntentWrapper;

public class ArrowSelector extends ImageSelectorBase<Arrow> {

    private static final int ARROW_REQUEST_CODE = 5;
    private static final int ARROW_ADD_REQUEST_CODE = 6;

    public ArrowSelector(Context context) {
        this(context, null);
    }

    public ArrowSelector(Context context, AttributeSet attrs) {
        super(context, attrs);
        defaultActivity = ArrowListActivity.class;
        requestCode = ARROW_REQUEST_CODE;
    }

    @NonNull
    @Override
    protected IntentWrapper getAddIntent() {
        return EditArrowFragment.createIntent()
                .forResult(ARROW_ADD_REQUEST_CODE);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @NonNull Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == Activity.RESULT_OK && requestCode == ARROW_ADD_REQUEST_CODE) {
            setItemId(null);
        }
    }

    public void setItemId(@Nullable Long arrowId) {
        Arrow item = null;
        if (arrowId != null) {
            item = Arrow.Companion.get(arrowId);
        }
        if (item == null) {
            List<Arrow> all = Arrow.Companion.getAll();
            if (all.size() > 0) {
                item = all.get(0);
            }
        }
        setItem(item);
    }
}

File: app/src/main/java/de/dreier/mytargets/views/selector/ArrowSelector.kt
/*
 * Copyright (C) 2018 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.views.selector

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.databinding.DataBindingUtil
import android.util.AttributeSet
import de.dreier.mytargets.R
import de.dreier.mytargets.app.ApplicationInstance
import de.dreier.mytargets.databinding.SelectorItemImageDetailsBinding
import de.dreier.mytargets.shared.models.db.Arrow

class ArrowSelector @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null
) : SelectorBase<Arrow>(context, attrs, R.layout.selector_item_image_details, ARROW_REQUEST_CODE) {

    private lateinit var binding: SelectorItemImageDetailsBinding

    private val arrowDAO = ApplicationInstance.db.arrowDAO()

    override fun bindView(item: Arrow) {
        binding = DataBindingUtil.bind(view)
        binding.name.text = item.name
        binding.image.setImageDrawable(item.thumbnail!!.roundDrawable)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK && requestCode == ARROW_ADD_REQUEST_CODE) {
            setItemId(null)
        }
    }

    fun setItemId(arrowId: Long?) {
        var item: Arrow? = null
        if (arrowId != null) {
            item = arrowDAO.loadArrowOrNull(arrowId)
        }
        if (item == null) {
            val all = arrowDAO.loadArrows()
            if (all.isNotEmpty()) {
                item = all[0]
            }
        }
        setItem(item)
    }

    companion object {
        const val ARROW_REQUEST_CODE = 5
        const val ARROW_ADD_REQUEST_CODE = 6
    }
}
